File: runtime/Go/src/antlr4/IntervalSet.go
package antlr4

import (
	"strings"
	"strconv"
	"math"
)

type Interval struct {
	start int
	stop int
}

/* stop is not included! */
func NewInterval(start, stop int) Interval{
	i := new(Interval)

	i.start = start
	i.stop = stop
	return i
}

func (i *Interval) contains(item int) {
	return item >= i.start && item < i.stop
}

func (i *Interval) toString() string {
	if(i.start==i.stop-1) {
		return strconv.Itoa(i.start)
	} else {
		return strconv.Itoa(i.start) + ".." + strconv.Itoa(i.stop-1)
	}
}

func (i *Interval) length() int {
	return i.stop - i.start
}

type IntervalSet struct {
	intervals []Interval
	readOnly bool
}

func NewIntervalSet() *IntervalSet {

	i := new(IntervalSet)

	i.intervals = nil
	i.readOnly = false

	return i
}

func (i *IntervalSet) first() int {
	if (len(i.intervals)==0) {
		return TokenInvalidType
	} else {
		return i.intervals[0].start
	}
}

func (i *IntervalSet) addOne(v int) {
	i.addInterval(NewInterval(v, v + 1))
}

func (i *IntervalSet) addRange(l, h int) {
	i.addInterval(NewInterval(l, h + 1))
}

func (is *IntervalSet) addInterval(v Interval) {
	if (is.intervals == nil) {
		is.intervals = make([]Interval, 0)
		is.intervals = append( is.intervals, v )
	} else {
		// find insert pos
		for k := 0; k < len(is.intervals); k++ {
			var i = is.intervals[k]
			// distinct range -> insert
			if (v.stop < i.start) {
				is.intervals.splice(k, 0, v)
				return
			} else if (v.stop == i.start) {
				is.intervals[k].start = v.start
				return
			} else if (v.start <= i.stop) {
				is.intervals[k] = NewInterval(math.Min(i.start, v.start), math.Max(i.stop, v.stop))
				is.reduce(k)
				return
			}
		}
		// greater than any existing
		is.intervals = append(is.intervals, v)
	}
}

func (i *IntervalSet) addSet(other IntervalSet) *IntervalSet {
	if (other.intervals != nil) {
		for k := 0; k < len(other.intervals); k++ {
			var i2 = other.intervals[k]
			i.addInterval(NewInterval(i2.start, i2.stop))
		}
	}
	return i
}

func (i *IntervalSet) reduce(k int) {
	// only need to reduce if k is not the last
	if (k < len(i.intervals) - 1) {
		var l = i.intervals[k]
		var r = i.intervals[k + 1]
		// if r contained in l
		if (l.stop >= r.stop) {
			i.intervals.pop(k + 1)
			i.reduce(k)
		} else if (l.stop >= r.start) {
			i.intervals[k] = NewInterval(l.start, r.stop)
			i.intervals.pop(k + 1)
		}
	}
}

func (is *IntervalSet) complement(start int, stop int) *IntervalSet {
    var result = NewIntervalSet()
    result.addInterval(NewInterval(start,stop+1))
    for i := 0; i< len(is.intervals); i++ {
        result.removeRange(is.intervals[i])
    }
    return result
}

func (i *IntervalSet) contains(item Interval) bool {
	if (i.intervals == nil) {
		return false
	} else {
		for k := 0; k < len(i.intervals); k++ {
			if(i.intervals[k].contains(item)) {
				return true
			}
		}
		return false
	}
}

func (is *IntervalSet) length() int {
	len := 0

	for _,v := range is.intervals {
		len += v.length()
	}

	return len
}

func (is *IntervalSet) removeRange(v Interval) {
    if v.start==v.stop-1 {
        is.removeOne(v.start)
    } else if (is.intervals!=nil) {
        k:= 0
        for n :=0; n<len( is.intervals ); n++ {
            var i = is.intervals[k]
            // intervals are ordered
            if (v.stop<=i.start) {
                return
            } else if(v.start>i.start && v.stop<i.stop) {
                is.intervals[k] = NewInterval(i.start, v.start)
                var x = NewInterval(v.stop, i.stop)
                is.intervals.splice(k, 0, x)
                return
            } else if(v.start<=i.start && v.stop>=i.stop) {
                is.intervals.splice(k, 1)
                k = k - 1 // need another pass
            } else if(v.start<i.stop) {
                is.intervals[k] = NewInterval(i.start, v.start)
            } else if(v.stop<i.stop) {
                is.intervals[k] = NewInterval(v.stop, i.stop)
            }
            k += 1
        }
    }
}

func (is *IntervalSet) removeOne(v *Interval) {
	if(v.start==v.stop-1) {
		is.removeOne(v.start)
	} else if (is.intervals!=nil) {
		var k = 0
		for n := 0; n < len(is.intervals); n++ {
			i := is.intervals[k]
			// intervals are ordered
			if v.stop <= i.start {
				return
			} else if v.start>i.start && v.stop<i.stop {
				// check for including range, split it
				is.intervals[k] = NewInterval(i.start, v.start)
				var x = NewInterval(v.stop, i.stop)
				is.intervals.splice(k, 0, x)
				return
			} else if(v.start<=i.start && v.stop>=i.stop) {
				is.intervals.splice(k, 1)
				k = k - 1; // need another pass
			} else if(v.start<i.stop) {
				is.intervals[k] = NewInterval(i.start, v.start)
			} else if(v.stop<i.stop) {
				is.intervals[k] = NewInterval(v.stop, i.stop)
			}
			k += 1
		}
	}
}

func (i *IntervalSet) toString() string {
	return i.toStringVerbose(nil,nil,false)
}

func (i *IntervalSet) toStringVerbose(literalNames []string, symbolicNames []string, elemsAreChar bool) string {

	if (i.intervals == nil) {
		return "{}"
	} else if(literalNames!=nil || symbolicNames!=nil) {
		return i.toTokenString(literalNames, symbolicNames)
	} else if(elemsAreChar) {
		return i.toCharString()
	} else {
		return i.toIndexString()
	}
}

func (is *IntervalSet) toCharString() {
	var names = make([]string, len(is.intervals))

	for i := 0; i < len( is.intervals ); i++ {
		var v = is.intervals[i]
		if(v.stop==v.start+1) {
			if ( v.start== TokenEOF ) {
				append(names, "<EOF>")
			} else {
				append(names, ("'" + string(v.start) + "'"))
			}
		} else {
			append(names, "'" + string(v.start) + "'..'" + string(v.stop-1) + "'")
		}
	}
	if (len(names) > 1) {
		return "{" + strings.Join(names, ", ") + "}"
	} else {
		return names[0]
	}
}


func (is *IntervalSet) toIndexString() {
	var names = make([]string, 0)
	for i := 0; i < len( is.intervals ); i++ {
		var v = is.intervals[i]
		if(v.stop==v.start+1) {
			if ( v.start==TokenEOF ) {
				names = append( names, "<EOF>")
			} else {
				names = append( names, string(v.start))
			}
		} else {
			names = append( names, string(v.start) + ".." + string(v.stop-1))
		}
	}
	if (len(names) > 1) {
		return "{" + strings.Join(names, ", ") + "}"
	} else {
		return names[0]
	}
}


func (is *IntervalSet) toTokenString(literalNames []string, symbolicNames []string) string {
	var names = make([]string, 0)
	for i := 0; i < len( is.intervals ); i++ {
		var v = is.intervals[i]
		for j := v.start; j < v.stop; j++ {
			names = append(names, is.elementName(literalNames, symbolicNames, j))
		}
	}
	if (len(names) > 1) {
		return "{" + strings.Join(names,", ") + "}"
	} else {
		return names[0]
	}
}

func (i *IntervalSet) elementName(literalNames []string, symbolicNames []string, a int) string {
	if (a == TokenEOF) {
		return "<EOF>"
	} else if (a == TokenEpsilon) {
		return "<EPSILON>"
	} else {
		return literalNames[a] || symbolicNames[a]
	}
}

File: runtime/JavaScript/src/antlr4/IntervalSet.js
/* Copyright (c) 2012-2017 The ANTLR Project. All rights reserved.
 * Use of this file is governed by the BSD 3-clause license that
 * can be found in the LICENSE.txt file in the project root.
 */

const {Token} = require('./Token');

/* stop is not included! */
class Interval {

	constructor(start, stop) {
		this.start = start;
		this.stop = stop;
	}

	clone() {
		return new Interval(this.start, this.stop);
	}

	contains(item) {
		return item >= this.start && item < this.stop;
	}

	toString() {
		if(this.start===this.stop-1) {
			return this.start.toString();
		} else {
			return this.start.toString() + ".." + (this.stop-1).toString();
		}
	}

	get length(){
		return this.stop - this.start;
	}
}


class IntervalSet {
	constructor() {
		this.intervals = null;
		this.readOnly = false;
	}

	first(v) {
		if (this.intervals === null || this.intervals.length===0) {
			return Token.INVALID_TYPE;
		} else {
			return this.intervals[0].start;
		}
	}

	addOne(v) {
		this.addInterval(new Interval(v, v + 1));
	}

	addRange(l, h) {
		this.addInterval(new Interval(l, h + 1));
	}

	addInterval(toAdd) {
		if (this.intervals === null) {
			this.intervals = [];
			this.intervals.push(toAdd.clone());
		} else {
			// find insert pos
			for (let pos = 0; pos < this.intervals.length; pos++) {
				const existing = this.intervals[pos];
				// distinct range -> insert
				if (toAdd.stop < existing.start) {
					this.intervals.splice(pos, 0, toAdd);
					return;
				}
				// contiguous range -> adjust
				else if (toAdd.stop === existing.start) {
					this.intervals[pos] = new Interval(toAdd.start, existing.stop)
					return;
				}
				// overlapping range -> adjust and reduce
				else if (toAdd.start <= existing.stop) {
					this.intervals[pos] = new Interval(Math.min(existing.start, toAdd.start), Math.max(existing.stop, toAdd.stop));
					this.reduce(pos);
					return;
				}
			}
			// greater than any existing
			this.intervals.push(toAdd.clone());
		}
	}

	addSet(other) {
		if (other.intervals !== null) {
			other.intervals.forEach( toAdd => this.addInterval(toAdd), this);
		}
		return this;
	}

	reduce(pos) {
		// only need to reduce if pos is not the last
		if (pos < this.intervals.length - 1) {
			const current = this.intervals[pos];
			const next = this.intervals[pos + 1];
			// if next contained in current
			if (current.stop >= next.stop) {
				this.intervals.splice(pos + 1, 1);
				this.reduce(pos);
			} else if (current.stop >= next.start) {
				this.intervals[pos] = new Interval(current.start, next.stop);
				this.intervals.splice(pos + 1, 1);
			}
		}
	}

	complement(start, stop) {
		const result = new IntervalSet();
		result.addInterval(new Interval(start, stop + 1));
		if(this.intervals !== null)
			this.intervals.forEach(toRemove => result.removeRange(toRemove));
		return result;
	}

	contains(item) {
		if (this.intervals === null) {
			return false;
		} else {
			for (let k = 0; k < this.intervals.length; k++) {
				if(this.intervals[k].contains(item)) {
					return true;
				}
			}
			return false;
		}
	}

	removeRange(toRemove) {
		if(toRemove.start===toRemove.stop-1) {
			this.removeOne(toRemove.start);
		} else if (this.intervals !== null) {
			let pos = 0;
			for(let n=0; n<this.intervals.length; n++) {
				const existing = this.intervals[pos];
				// intervals are ordered
				if (toRemove.stop<=existing.start) {
					return;
				}
				// check for including range, split it
				else if(toRemove.start>existing.start && toRemove.stop<existing.stop) {
					this.intervals[pos] = new Interval(existing.start, toRemove.start);
					const x = new Interval(toRemove.stop, existing.stop);
					this.intervals.splice(pos, 0, x);
					return;
				}
				// check for included range, remove it
				else if(toRemove.start<=existing.start && toRemove.stop>=existing.stop) {
					this.intervals.splice(pos, 1);
					pos = pos - 1; // need another pass
				}
				// check for lower boundary
				else if(toRemove.start<existing.stop) {
					this.intervals[pos] = new Interval(existing.start, toRemove.start);
				}
				// check for upper boundary
				else if(toRemove.stop<existing.stop) {
					this.intervals[pos] = new Interval(toRemove.stop, existing.stop);
				}
				pos += 1;
			}
		}
	}

	removeOne(value) {
		if (this.intervals !== null) {
			for (let i = 0; i < this.intervals.length; i++) {
				const existing = this.intervals[i];
				// intervals are ordered
				if (value < existing.start) {
					return;
				}
				// check for single value range
				else if (value === existing.start && value === existing.stop - 1) {
					this.intervals.splice(i, 1);
					return;
				}
				// check for lower boundary
				else if (value === existing.start) {
					this.intervals[i] = new Interval(existing.start + 1, existing.stop);
					return;
				}
				// check for upper boundary
				else if (value === existing.stop - 1) {
					this.intervals[i] = new Interval(existing.start, existing.stop - 1);
					return;
				}
				// split existing range
				else if (value < existing.stop - 1) {
					const replace = new Interval(existing.start, value);
					existing.start = value + 1;
					this.intervals.splice(i, 0, replace);
					return;
				}
			}
		}
	}

	toString(literalNames, symbolicNames, elemsAreChar) {
		literalNames = literalNames || null;
		symbolicNames = symbolicNames || null;
		elemsAreChar = elemsAreChar || false;
		if (this.intervals === null) {
			return "{}";
		} else if(literalNames!==null || symbolicNames!==null) {
			return this.toTokenString(literalNames, symbolicNames);
		} else if(elemsAreChar) {
			return this.toCharString();
		} else {
			return this.toIndexString();
		}
	}

	toCharString() {
		const names = [];
		for (let i = 0; i < this.intervals.length; i++) {
			const existing = this.intervals[i];
			if(existing.stop===existing.start+1) {
				if ( existing.start===Token.EOF ) {
					names.push("<EOF>");
				} else {
					names.push("'" + String.fromCharCode(existing.start) + "'");
				}
			} else {
				names.push("'" + String.fromCharCode(existing.start) + "'..'" + String.fromCharCode(existing.stop-1) + "'");
			}
		}
		if (names.length > 1) {
			return "{" + names.join(", ") + "}";
		} else {
			return names[0];
		}
	}

	toIndexString() {
		const names = [];
		for (let i = 0; i < this.intervals.length; i++) {
			const existing = this.intervals[i];
			if(existing.stop===existing.start+1) {
				if ( existing.start===Token.EOF ) {
					names.push("<EOF>");
				} else {
					names.push(existing.start.toString());
				}
			} else {
				names.push(existing.start.toString() + ".." + (existing.stop-1).toString());
			}
		}
		if (names.length > 1) {
			return "{" + names.join(", ") + "}";
		} else {
			return names[0];
		}
	}

	toTokenString(literalNames, symbolicNames) {
		const names = [];
		for (let i = 0; i < this.intervals.length; i++) {
			const existing = this.intervals[i];
			for (let j = existing.start; j < existing.stop; j++) {
				names.push(this.elementName(literalNames, symbolicNames, j));
			}
		}
		if (names.length > 1) {
			return "{" + names.join(", ") + "}";
		} else {
			return names[0];
		}
	}

	elementName(literalNames, symbolicNames, token) {
		if (token === Token.EOF) {
			return "<EOF>";
		} else if (token === Token.EPSILON) {
			return "<EPSILON>";
		} else {
			return literalNames[token] || symbolicNames[token];
		}
	}

	get length(){
		return this.intervals.map( interval => interval.length ).reduce((acc, val) => acc + val);
	}
}

module.exports = {
	Interval,
	IntervalSet
};
